src/ray/object_manager/pull_manager.cc

```diff
@@ -399,7 +402,10 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
   it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());
   it->second.spilled_url = spilled_url;
   it->second.spilled_node_id = spilled_node_id;
-  if (!it->second.object_size_set) {
+  it->second.pending_object_creation = pending_creation;
+  if (!it->second.object_size_set && !pending_creation) {
```

Comment:
This is the key change. This was causing us to start trying to pull the object even though it was still pending creation. It would also put us way over the pull capacity because we thought the object had size 0.

Reply:
Have we added a unit test for this case?

Reply:
Btw, When does the object size set although it is pending creation? Is it impossible to unify them to one var?

Reply:
Actually I ended up reverting this because it caused a different issue when there's a failure.

Eventually, we should clean this up and put them into one flag to indicate that we're waiting for object metadata, but for now I think this is the simplest.

Reply:
Unfortunately, I can't easily add a unit test for this case because it involves multiple classes. But the Python reconstruction test caught this issue.

Reply:
Never mind, was able to add a unit test to the OBOD for the fix!